Nuked TGenotype

cseed opened the pull request:
TGenotype gone and the Genotype class is gone.  The few remaining legacy genotypes are stored generically as annotations.  The changes are primarily what's naturally forced.

Main unforced change was to revise the way VSM.gen apportions sizes.  Time seems roughly the same and the new generator turned up a few bugs so I think it is better.

@tpoterba Sorry, this is a monster.  I can try to break it up if needed.

tpoterba commented on python/hail/dataset.py:
```diff
@@ -1664,11 +1643,9 @@ def export_vcf(self, output, append_to_header=None, parallel=False):
-        :py:class:`~hail.expr.TGenotype`, then the FORMAT fields will be GT, AD, DP, GQ, and PL.
-        If the type is :py:class:`~hail.expr.TStruct`,
-        then the exported FORMAT fields will be the names of each field of the Struct.
+        The FORMAT field is generated from the genotype type, which
```
funky language here, something got deleted and not clarified.

tpoterba commented on python/hail/dataset.py:
```diff
@@ -2152,7 +2130,7 @@ def filter_samples_expr(self, expr, keep=True):
         - ``sa``: sample annotations
         - ``global``: global annotations
-        - ``gs`` (*Aggregable[Genotype]*): aggregable of :ref:`genotype` for sample ``s``
+        - ``gs`` (*Aggregable[Genotype]*): aggregable of genotype for sample ``s``
```
aggregable[genotype] is wrong

tpoterba replied:
this is found a bunch of places, actually.

tpoterba replied:
Maybe 
```
- ``gs`` (*Aggregable[Struct]*): aggregable of genotype structs for sample ``s``
```

cseed replied:
It doesn't have to be a struct.  I'm trying to sloppily use "genotype" to mean whatever the cell in the matrix is so I don't have to revise all the documentation when it is going to get revised again for the new python API.  Another idea: I could still write `Aggregable[genotype]`, but have the genotype link out to `genotype_schema`.

tpoterba replied:
Alright, yeah. Don't worry about this.

tpoterba commented on python/hail/docs/overview.rst:
```diff
@@ -19,7 +19,7 @@ Variant Dataset (VDS)
 Hail represents a genetic data set as a matrix where the rows are keyed by
 :ref:`variant(gr)` objects, the columns are keyed by samples, and each cell is a
-:ref:`genotype` object. :ref:`variant(gr)` objects and :ref:`genotype` objects each
+genotype object. :ref:`variant(gr)` objects and genotype objects each
```
genotype object is now an outdated term

tpoterba commented on python/hail/representation/genotype.py:
```diff
@@ -346,6 +34,12 @@ def __str__(self):
         return 'Call(gt=%s)' % self._gt
 
+    def __eq__(self, other):
+        return self.gt == other.gt
```
`return isinstance(other, Call) and self.gt == other.gt`

tpoterba commented on python/hail/representation/genotype.py:
```diff
@@ -346,6 +34,12 @@ def __str__(self):
+        return self.gt == other.gt
+
+    def __hash__(self):
+        return hash(self.gt)
```
can we add some magic number here so hash(Call(1)) isn't hash(1)?

tpoterba commented on python/hail/tests/tests.py:
```diff
@@ -329,7 +316,8 @@ def test_dataset(self):
         sample2.split_multi().impute_sex().variant_schema
 
-        self.assertTrue(isinstance(sample2.genotype_schema, TGenotype))
+        # FIXME htsGenotypeSchema
```
can you explain this a bit?

cseed replied:
Yeah, I wanted to check the genotype_schema is hts_genotype_schema, but I need to export it to python.  Will do.

tpoterba commented on python/hail2/tests.py:
```diff
@@ -170,17 +170,17 @@ def test_select(self):
     def test_aggregate(self):
-        schema = TStruct(['status', 'gt', 'qPheno'],
-                         [TInt32(), TGenotype(), TInt32()])
+        schema = TStruct(['status', 'GT', 'qPheno'],
```
the py-expr-2 rebase is going to be very real.

tpoterba commented on src/main/scala/is/hail/methods/ImputeSexPlink.scala:
```diff
@@ -17,19 +17,27 @@ object ImputeSexPlink {
       case TVariant(x, _) => x.asInstanceOf[GenomeReference]
     }
 
+    vsm.typecheck()
```
seems like you meant to remove these?

tpoterba commented on src/main/scala/is/hail/stats/CallStatsCombiner.scala:
```diff
@@ -15,18 +15,19 @@ object CallStats {
 
 case class CallStats(alleleCount: IndexedSeq[Int], alleleFrequency: Option[IndexedSeq[Double]], alleleNumber: Int,
   genotypeCount: IndexedSeq[Int]) {
+  println(alleleNumber, genotypeCount.toSeq, genotypeCount.sum)
```
remove

tpoterba commented on src/main/scala/is/hail/variant/Genotype.scala:
```diff
@@ -674,20 +345,25 @@ object Genotype {
+        dp.map(_ + ad.map(_.sum).getOrElse(0)).orNull,
+        gq.orNull,
+        pl.map(a => a: IndexedSeq[Int]).orNull)
+      // FIXME
```
?

tpoterba commented on src/main/scala/is/hail/variant/HTSGenotypeView.scala:
```diff
@@ -8,17 +8,13 @@ import is.hail.utils._
 }
 
-// FIXME: This is removed, and StructGenotypeView becomes the only genotype view when TGenotype is removed
+// FIXME remove
```
do you want to do this cleanup in a followup PR?

tpoterba commented on src/main/scala/is/hail/variant/VariantDataset.scala:
```diff
@@ -290,51 +282,11 @@ class VariantDatasetFunctions(private val vsm: VariantSampleMatrix) extends AnyV
   def filterAlleles(filterExpr: String, variantExpr: String = "",
     keep: Boolean = true, subset: Boolean = true, leftAligned: Boolean = false, keepStar: Boolean = false): VariantSampleMatrix = {
+    if (!vsm.genotypeSignature.isOfType(Genotype.htsGenotypeType))
+      fatal(s"filter_alleles: genotype_schema must be TGenotype or the HTS genotype schema, found: ${ vsm.genotypeSignature }")
```
tgenotype reference

tpoterba commented on src/test/resources/filter.vcf:
```diff
@@ -1,11 +1,9 @@
 ##fileformat=VCFv4.1
+##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
```
changes to this file are to make it compatible with the HTS schema?

tpoterba replied:
to this and all the following files, rather.

cseed replied:
Yes, exactly.

cseed commented:
I addressed some comments.  I still need to:
 - expose hts_genotype_schema in python, and
 - figure out what to rewrite instead of "genotype" in the VariantDataset docs.

cseed commented:
I pushed more changes to remove is{Not}Called and renamed isCalledNonRef => isNonRef.

I will address comments and rebase after the rest of the stack goes in.

cseed commented:
Added `Type.hts_schema` in Python.  I'm not sure that's the right place for it and I couldn't make it a property because classes have properties (although it might be possible to do it with a property on the metaclass).

tpoterba commented:
OK, will take another look.

tpoterba commented on src/main/scala/is/hail/expr/Relational.scala:
```diff
@@ -185,11 +185,11 @@ object MatrixValue {
           it.map { case (v, (va, gs)) =>
+            val region = MemoryBuffer()
+            val rvb = new RegionValueBuilder(region)
+            val rv = RegionValue(region)
```
why this change? Wasn't it working with mutation?

cseed replied:
Didn't mean to put this in, I'll remove.

cseed replied:
Fixed.

tpoterba commented on src/main/scala/is/hail/variant/Genotype.scala:
```diff
@@ -54,71 +49,6 @@ class GTPair(val p: Int) extends AnyVal {
-}
-
 object Genotype {
   val htsGenotypeType: TStruct = TStruct(
```
This doesn't seem like the long-term design we want, because in order to run variantQC, you don't need something that is exactly GT/AD/DP/GQ/PL, but just something that contains those fields in any order (and could contain others). So we wouldn't want to check equality, but instead to check membership. Is this a temporary workaround to keep stuff working with the tests?

cseed replied:
This isn't used in variantQC.  It used in split_multi and related routines.

split_multi requires exactly this schema (although in an arbitrary order is fine), because it has a fixed schema on output and will delete any other genotype fields present.  Although the input can be in any order, the output is in the standard order.  I have to do this because we can't put an insertion inside a let and I need, for example, to compute GQ  from PL (without recording PL twice).  Also, instantiating the entire struct is way faster than doing 5 insertions (although that's a short term concern).

So ... I could make split_multi require the right fields but in any order.

tpoterba replied:
Alright. Let's revisit this again in a month or so, to evaluate whether it's still the right thing. Seems fine for now.

cseed commented:
Rebased.  Despite some linger issues (docs, handling the HTS schema) I think this is ready for another look.

tpoterba commented:
I'm not too concerned about docs, since it's all going to be revisited soon.

tpoterba commented on python/hail/tests/tests.py:
```diff
@@ -33,6 +33,14 @@ def float_eq(x, y, tol=10 ** -6):
 
 class ContextTests(unittest.TestCase):
+    # FIXME remove when attributes get nuked
+    def assert_equal_struct(self, s1, s2):
```
wait, hold off. Attributes have been nuked.

cseed replied:
Fixed!

cseed commented:
Python tests fixed.
